resolwe/flow/models/utils/duplicate.py

```diff
@@ -0,0 +1,301 @@
+                continue
+
+            setattr(new_obj, field_name, pk_mapping[value])
+
+        new_obj.pk = None
+        new_objects.append(new_obj)
+
+    return obj_class.objects.bulk_create(new_objects)
+
+
+def _rewire_inputs(data, pk_mapping):
+    """Rewire inputs of provided data objects."""
+    for datum in data:
+        for field_schema, fields in iterate_fields(
+            datum.input, datum.process.input_schema
+        ):
+            name = field_schema["name"]
+            value = fields[name]
+
+            if field_schema["type"].startswith("data:") and value in pk_mapping:
+                fields[name] = pk_mapping[value]
+
+            elif field_schema["type"].startswith("list:data:"):
+                fields[name] = [
+                    pk_mapping[pk] if pk in pk_mapping else pk for pk in value
+                ]
+
+
+def process_entity(inherit_collection, collection_mapping):
+    def processor(entity):
+        if inherit_collection:
+            if entity.collection.id in collection_mapping:
+                entity.collection = collection_mapping[entity.collection.pk]
+        else:
+            entity.collection = None
+
+        return entity
+
+    return processor
+
+
+def process_data(
+    inherit_collection, inherit_entity, collection_mapping, entity_mapping
+):
+    def processor(datum):
+        if inherit_collection:
+            if datum.collection.id in collection_mapping:
+                datum.collection = collection_mapping[datum.collection.pk]
+        else:
+            datum.collection = None
+
+        if inherit_entity:
+            if datum.entity.id in entity_mapping:
+                datum.entity = entity_mapping[datum.entity.pk]
+        else:
+            datum.entity = None
+
+        return datum
+
+    return processor
+
+
+def copy_objects(objects, contributor, name_prefix, obj_processor=None):
+    """"""
+    first = objects.first()
+    if not first:
+        return objects
+
+    name_max_length = first._meta.get_field('name').max_length
+    content_type = ContentType.objects.get_for_model(first)
+    model = first._meta.model
+
+    new_objects = []
+    for obj in objects:
+        new_obj = deepcopy(obj)
+        new_obj.pk = None
+        new_obj.slug = None
+        new_obj.contributor = contributor
+        new_obj.name = "{} {}".format(name_prefix, obj.name)
+
+        if len(new_obj.name) > name_max_length:
+            new_obj.name = "{}...".format(new_obj.name[: name_max_length - 3])
+
+        if obj_processor:
+            new_obj = obj_processor(new_obj)
+
+        new_objects.append(new_obj)
+
+    try:
+        # Add another atomic block to avoid corupting the main one.
+        with transaction.atomic():
+            model.objects.bulk_create(new_objects)
+    except IntegrityError:
+        # Probably a slug collision occured, try to create objects one by one.
+        for obj in new_objects:
+            obj.slug = None
```

Comment:
This will probably lead to the same error as the code above?
Sooner or later the same slug clash will happen as it did above?

Reply:
Slug is generated based only on slugs already in the database. This means that the second conflicting slug will be changed if the first one is already written in the database.